src/comments.js

```diff
@@ -129,40 +140,60 @@ function attach(comments, ast, text) {
   comments.forEach(function(comment) {
     decorateComment(ast, comment, text);
 
-    var pn = comment.precedingNode;
-    var en = comment.enclosingNode;
-    var fn = comment.followingNode;
-
-    if (pn && fn) {
-      var tieCount = tiesToBreak.length;
-      if (tieCount > 0) {
-        var lastTie = tiesToBreak[tieCount - 1];
-
-        assert.strictEqual(
-          lastTie.precedingNode === comment.precedingNode,
-          lastTie.followingNode === comment.followingNode
-        );
+    const pn = comment.precedingNode;
+    const globalPn = comment.globalPrecedingNode;
+    const en = comment.enclosingNode;
+    const fn = comment.followingNode;
+    const isStartOfFile = comment.loc.start.line === 1;
+
+    if (util.hasNewline(text, locStart(comment), true) || isStartOfFile) {
+      // If a comment exists on its own line, prefer a leading comment.
+      // We also need to check if it's the first line of the file.
+
+      if (fn) {
+        // Always a leading comment.
+        addLeadingComment(fn, comment);
+      } else if (pn) {
+        addTrailingComment(pn, comment);
+      } else if (en) {
+        addDanglingComment(en, comment);
+      } else {
+        // TODO: If there are no nodes at all, we should still somehow
+        // print the comment.
+      }
 
-        if (lastTie.followingNode !== comment.followingNode) {
-          breakTies(tiesToBreak, text);
-        }
+    } else if(util.hasNewline(text, locEnd(comment))) {
+      // There is content before this comment on the same line, so if
+      // there's a newline after it, prefer a trailing comment. A
+      // trailing comment *always* attaches itself to the previous
+      // node, no matter if there's any syntax between them.
+      if (pn || globalPn) {
+        // Always a trailing comment
+        addTrailingComment(pn || globalPn, comment);
       }
 
-      tiesToBreak.push(comment);
-    } else if (pn) {
-      // No contest: we have a trailing comment.
-      breakTies(tiesToBreak, text);
-      addTrailingComment(pn, comment);
-    } else if (fn) {
-      // No contest: we have a leading comment.
-      breakTies(tiesToBreak, text);
-      addLeadingComment(fn, comment);
-    } else if (en) {
-      // The enclosing node has no child nodes at all, so what we
-      // have here is a dangling comment, e.g. [/* crickets */].
-      breakTies(tiesToBreak, text);
-      addDanglingComment(en, comment);
     } else {
+      // Otherwise, text exists both before and after the comment on
```

Comment:
The works fine, the output is `call(/* comment */ true);`.

The tie breaking algorithm does work correctly with that. The way it works is it looks on each side of the comment and will attach to whichever node it finds that *only* has whitespace between the node and the comment. The above code has a `(` between the `call` node and the comment, and sees there is nothing between the comment and `true`, so it picks `true`.

I got that algorithm from recast and it's pretty smart, but we only want to use it as a last resort.

You're right though, I should update the comment that this covers that case too.